src/qt/ee_debugwindow.cpp

```diff
@@ -0,0 +1,1231 @@
+        }
+    }
+    menu.exec(QCursor::pos());
+}
+
+void EEDebugWindow::register_context_menu(const QPoint &qpos)
+{
+    if(emulation_running()) return;
+    int32_t row = register_table->currentRow();
+    QMenu menu;
+    if(current_cpu == DebuggerCpu::EE)
+    {
+        auto reg_def = ee_registers[row - 1]; // hack for pc
+        auto bps = ee_breakpoint_list->get_breakpoints_at_register(reg_def);
+
+
+        menu.addAction(tr("Add Breakpoint..."), this, [&]{
+            breakpoint_window bpw;
+            DebuggerBreakpoint* new_bp = bpw.make_register_breakpoint(reg_def);
+            if(new_bp)
+                add_breakpoint(new_bp);
+        });
+
+        menu.addAction(tr("Add new Register AND breakpoint..."), this, [&]{
+            breakpoint_window bpw;
+            DebuggerBreakpoint* reg_bp = bpw.make_register_breakpoint(reg_def);
+            if(!reg_bp) return;
+            AndBreakpointWindow and_bpw;
+            DebuggerBreakpoint* and_bp = and_bpw.make_and_breakpoint(&ee_registers, reg_bp);
+            if(and_bp)
+                add_breakpoint(and_bp);
+        });
+
+        if(!bps.empty())
+        {
+            menu.addSeparator();
+            for(uint64_t bp : bps)
+            {
+                QString menu_item_name = "Delete " + QString(ee_breakpoint_list->get_breakpoints()[bp]->get_description().c_str());
+                menu.addAction(menu_item_name, this, [=]{delete_breakpoint_by_index(bp);});
+            }
+        }
+    }
+    else
+    {
+        auto reg_def = iop_registers[row - 1]; // hack for pc
+        auto bps = iop_breakpoint_list->get_breakpoints_at_register(reg_def);
+
+        menu.addAction(tr("Add Breakpoint..."), this, [&]{
+            breakpoint_window bpw;
+            DebuggerBreakpoint* new_bp = bpw.make_register_breakpoint(reg_def);
+            if(new_bp)
+                add_breakpoint(new_bp);
+        });
+
+        menu.addAction(tr("Add new Register AND breakpoint..."), this, [&]{
+            breakpoint_window bpw;
+            DebuggerBreakpoint* reg_bp = bpw.make_register_breakpoint(reg_def);
+            if(!reg_bp) return;
+            AndBreakpointWindow and_bpw;
+            DebuggerBreakpoint* and_bp = and_bpw.make_and_breakpoint(&iop_registers, reg_bp);
+            if(and_bp)
+                add_breakpoint(and_bp);
+        });
+
+        if(!bps.empty())
+        {
+            menu.addSeparator();
+            for(uint64_t bp : bps)
+            {
+                QString menu_item_name = "Delete " + QString(iop_breakpoint_list->get_breakpoints()[bp]->get_description().c_str());
+                menu.addAction(menu_item_name, this, [=]{delete_breakpoint_by_index(bp);});
+            }
+        }
+    }
+
+    menu.exec(QCursor::pos());
+}
+
+EEDebugWindow::~EEDebugWindow()
+{
+
+}
+
+void EEDebugWindow::on_disassembly_up_button_clicked()
+{
+    get_current_cpu_disassembly_location() -= 2048;
```

Comment:
Having a function which returns a reference to a reference like this in order to update it in this fashion is confusing, and prone to creating bugs a setter can handle.

A better solution would be to have a setter which has logic that detects invalid memory regions and the such, which could be called like `set_current_cpu_disassembly_location(get_current_cpu_disassembly_location() - 2048);`

An even more ideal solution would be to stream in disassembly or memory into the debug window. For example, 2KB worth of disassembly could be loaded, and when the user scrolls up .5 from the midpoint, the bottom 1KB could be pruned and and additional 1KB of disassembly could be appended up top.